pkg/storage/tsdb/upload.go

```diff
@@ -0,0 +1,127 @@
+	if err != nil {
+		return errors.Wrap(err, "gather meta file stats")
+	}
+
+	metaEncoded := strings.Builder{}
+	if err := meta.Write(&metaEncoded); err != nil {
+		return errors.Wrap(err, "encode meta file")
+	}
+
+	if err := objstore.UploadDir(ctx, logger, bkt, filepath.Join(blockDir, block.ChunksDirname), path.Join(id.String(), block.ChunksDirname)); err != nil {
+		return cleanUp(logger, bkt, id, errors.Wrap(err, "upload chunks"))
+	}
+
+	if err := objstore.UploadFile(ctx, logger, bkt, filepath.Join(blockDir, block.IndexFilename), path.Join(id.String(), block.IndexFilename)); err != nil {
+		return cleanUp(logger, bkt, id, errors.Wrap(err, "upload index"))
+	}
+
+	// Meta.json always need to be uploaded as a last item. This will allow to assume block directories without meta file to be pending uploads.
+	if err := bkt.Upload(ctx, path.Join(id.String(), block.MetaFilename), strings.NewReader(metaEncoded.String())); err != nil {
+		// Don't call cleanUp here. Despite getting error, meta.json may have been uploaded in certain cases,
+		// and even though cleanUp will not see it yet, meta.json may appear in the bucket later.
+		// (Eg. S3 is known to behave this way when it returns 503 "SlowDown" error).
+		// If meta.json is not uploaded, this will produce partial blocks, but such blocks will be cleaned later.
+		return errors.Wrap(err, "upload meta file")
+	}
+
+	return nil
+}
+
+// Hopefully can be replaced with Thanos public function: https://github.com/thanos-io/thanos/pull/5400
+func gatherFileStats(blockDir string) (res []metadata.File, _ error) {
+	files, err := ioutil.ReadDir(filepath.Join(blockDir, block.ChunksDirname))
+	if err != nil {
+		return nil, errors.Wrapf(err, "read dir %v", filepath.Join(blockDir, block.ChunksDirname))
+	}
+	for _, f := range files {
+		mf := metadata.File{
+			RelPath:   filepath.Join(block.ChunksDirname, f.Name()),
+			SizeBytes: f.Size(),
+		}
+		res = append(res, mf)
+	}
+
+	indexFile, err := os.Stat(filepath.Join(blockDir, block.IndexFilename))
+	if err != nil {
+		return nil, errors.Wrapf(err, "stat %v", filepath.Join(blockDir, block.IndexFilename))
+	}
+	mf := metadata.File{
+		RelPath:   indexFile.Name(),
+		SizeBytes: indexFile.Size(),
+	}
+	res = append(res, mf)
+
+	metaFile, err := os.Stat(filepath.Join(blockDir, block.MetaFilename))
+	if err != nil {
+		return nil, errors.Wrapf(err, "stat %v", filepath.Join(blockDir, block.MetaFilename))
+	}
+	res = append(res, metadata.File{RelPath: metaFile.Name()})
+
+	sort.Slice(res, func(i, j int) bool {
+		return strings.Compare(res[i].RelPath, res[j].RelPath) < 0
+	})
+	return res, err
+}
+
+func cleanUp(logger log.Logger, bkt objstore.Bucket, id ulid.ULID, err error) error {
+	// Cleanup the dir with an uncancelable context.
+	cleanErr := block.Delete(context.Background(), logger, bkt, id)
+	if cleanErr != nil {
+		return errors.Wrapf(err, "failed to clean block after upload issue. Partial block in system. Err: %s", err.Error())
```

Comment:
doesn't that duplicate the error message?
It wraps the original error, but also adds the original error's message into the formatting string.

Reply:
ah, i think it's supposed to wrap `cleanErr` instead of `err`, right?

Reply:
```suggestion
		return errors.Wrapf(cleanErr, "failed to clean block after upload issue. Partial block in system. Err: %s", err.Error())
```